include/dsn/tool-api/network.h

```diff
@@ -275,7 +281,10 @@ class rpc_session : public ref_counter
     volatile session_state _connect_state;
 
     bool negotiation_succeed = false;
-    // TODO(zlw): add send pending message
+    // when the negotiation of a session isn't succeed,
+    // all messages are queued in _pending_messages.
+    // after connected, all of them are moved to "_messages"
+    std::vector<message_ex *> _pending_messages;
```

Comment:
Why need this pending queue of messages? The clients can buffer its messages before the negotiation completes. I think this is an over-optimzation.